wurst/network/Network.wurst

```diff
@@ -0,0 +1,922 @@
+				// network.getData() is locked in a WRITE-ONLY MODE
+				val stream = network.getData()
+				stream.writeInteger(...)
+				stream.writeReal(...)
+				stream.writeString(...)
+				...
+
+		3. To start sending the data, simply call:
+			network.start()			
+			// this will start the transmission
+			// keep in mind, that after you call this, you MUST NOT write/read to network.getData()
+			// as it is locked while the data is syncing
+		
+		4. To receive the data, specify the callback and read the data inside it:
+			network.onFinish((stream) -> begin
+				// here, network.getData() is locked in a READ-ONLY MODE and
+				// you can only READ from it
+				var int = stream.readInteger()
+				var real = stream.readReal()
+				var string = stream.readString()
+				...
+			end)
+
+			WARNING: Network is a one-time use class, and is automatically
+			destroyed upon finishing. If you need to send another transmission,
+			create a new one.
+			You are also expected to read all data immediately after receiving it,
+			as the DataStream is destroyed together with Network. 
+
+		5. When writing and reading, keep in mind that you MUST read data from
+		the stream in the same order that it was written, for each independent type.
+		That means that integers have to be written/read in the same order, same for
+		booleans, strings and reals.
+		If you need data of non-fixed length, use a static packet structure
+		with length fields, like so:
+			var field1 = stream.readInteger()
+			var field2 = stream.readInteger()
+			var length = stream.readInteger()
+			for i = 0 to length - 1
+				someArray[i] = stream.readInteger()
+
+		6. Alternatively, your classes can implement Serializable, like so:
+			class MyClass implements Serializable
+				integer field1 = ...
+				integer field2 = ...
+				integer field3 = ...
+				string field4 = ...
+
+				// optional stream-constructor
+				construct (DataStream stream)
+					deserialize(stream)
+
+				function serialize(DataStream stream)
+					stream.writeInteger(field1)
+					stream.writeInteger(field2)
+					stream.writeInteger(field3)
+					stream.writeString(field4)
+				
+				function deserialize(DataStream stream)
+					field1 = stream.readInteger()
+					field2 = stream.readInteger()
+					field3 = stream.readInteger()
+					field4 = stream.readString()
+
+		And then use it like so:
+			// writing
+			val myClass = ...
+			stream.write(myClass)
+			// reading (variant 1)
+			val myClass = ...
+			stream.read(myClass)
+			// reading (variant 2 with optional constructor)
+			val myClass = new MyClass(stream)
+
+	Implementation overview:
+		There are 3 core classes here:
+			DataStream
+			GamecacheStream
+			Network
+
+		DataStream is a hashtable-based container with stream semantics for writing
+		integers, reals, booleans, strings and serializables.
+		Because we want to keep gamecache keys below a certain size, we can't store
+		everything immediately in the gamecache without risking exhausting available keys.
```

Comment:
You mean that `gamecache` might have a collision? Source?

Reply:
Network performance degrades significantly with longer gamecache keys, since keys (as well as gamecache names themselves) also get sent over the network. This has been demonstrated by TriggerHappy previously in his findings for the Sync library.

https://www.hiveworkshop.com/pastebin/3a4f7861cb884e21312168bd654330585801/

Reply:
Is this documented in JassDoc? @lep ?

Code needs a direct reference if you're going to both (1) make claims about perf, and (2) they're actually relevant